File: src/java/main/org/apache/zookeeper/server/persistence/FileTxnSnapLog.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.zookeeper.server.persistence;

import java.io.File;
import java.io.IOException;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import org.apache.jute.Record;
import org.apache.zookeeper.KeeperException;
import org.apache.zookeeper.KeeperException.Code;
import org.apache.zookeeper.ZooDefs.OpCode;
import org.apache.zookeeper.server.DataTree;
import org.apache.zookeeper.server.DataTree.ProcessTxnResult;
import org.apache.zookeeper.server.Request;
import org.apache.zookeeper.server.ZooTrace;
import org.apache.zookeeper.server.persistence.TxnLog.TxnIterator;
import org.apache.zookeeper.txn.CreateSessionTxn;
import org.apache.zookeeper.txn.TxnHeader;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * This is a helper class 
 * above the implementations 
 * of txnlog and snapshot 
 * classes
 */
public class FileTxnSnapLog {
    //the direcotry containing the 
    //the transaction logs
    private final File dataDir;
    //the directory containing the
    //the snapshot directory
    private final File snapDir;
    private TxnLog txnLog;
    private SnapShot snapLog;
    public final static int VERSION = 2;
    public final static String version = "version-";
    
    private static final Logger LOG = LoggerFactory.getLogger(FileTxnSnapLog.class);
    
    /**
     * This listener helps
     * the external apis calling
     * restore to gather information
     * while the data is being 
     * restored.
     */
    public interface PlayBackListener {
        void onTxnLoaded(TxnHeader hdr, Record rec);
    }
    
    /**
     * the constructor which takes the datadir and 
     * snapdir.
     * @param dataDir the trasaction directory
     * @param snapDir the snapshot directory
     */
    public FileTxnSnapLog(File dataDir, File snapDir) throws IOException {
        LOG.debug("Opening datadir:{} snapDir:{}", dataDir, snapDir);

        this.dataDir = new File(dataDir, version + VERSION);
        this.snapDir = new File(snapDir, version + VERSION);
        if (!this.dataDir.exists()) {
            if (!this.dataDir.mkdirs()) {
                throw new IOException("Unable to create data directory "
                        + this.dataDir);
            }
        }
        if (!this.snapDir.exists()) {
            if (!this.snapDir.mkdirs()) {
                throw new IOException("Unable to create snap directory "
                        + this.snapDir);
            }
        }
        txnLog = new FileTxnLog(this.dataDir);
        snapLog = new FileSnap(this.snapDir);
    }
    
    /**
     * get the datadir used by this filetxn
     * snap log
     * @return the data dir
     */
    public File getDataDir() {
        return this.dataDir;
    }
    
    /**
     * get the snap dir used by this 
     * filetxn snap log
     * @return the snap dir
     */
    public File getSnapDir() {
        return this.snapDir;
    }
    
    /**
     * this function restores the server 
     * database after reading from the 
     * snapshots and transaction logs
     * @param dt the datatree to be restored
     * @param sessions the sessions to be restored
     * @param listener the playback listener to run on the 
     * database restoration
     * @return the highest zxid restored
     * @throws IOException
     */
    public long restore(DataTree dt, Map<Long, Integer> sessions, 
            PlayBackListener listener) throws IOException {
        snapLog.deserialize(dt, sessions);
        FileTxnLog txnLog = new FileTxnLog(dataDir);
        TxnIterator itr = txnLog.read(dt.lastProcessedZxid+1);
        long highestZxid = dt.lastProcessedZxid;
        TxnHeader hdr;
        while (true) {
            // iterator points to 
            // the first valid txn when initialized
            hdr = itr.getHeader();
            if (hdr == null) {
                //empty logs 
                return dt.lastProcessedZxid;
            }
            if (hdr.getZxid() < highestZxid && highestZxid != 0) {
                LOG.error(highestZxid + "(higestZxid) > "
                        + hdr.getZxid() + "(next log) for type "
                        + hdr.getType());
            } else {
                highestZxid = hdr.getZxid();
            }
            try {
                processTransaction(hdr,dt,sessions, itr.getTxn());
            } catch(KeeperException.NoNodeException e) {
               throw new IOException("Failed to process transaction type: " +
                     hdr.getType() + " error: " + e.getMessage(), e);
            }
            listener.onTxnLoaded(hdr, itr.getTxn());
            if (!itr.next()) 
                break;
        }
        return highestZxid;
    }
    
    /**
     * process the transaction on the datatree
     * @param hdr the hdr of the transaction
     * @param dt the datatree to apply transaction to
     * @param sessions the sessions to be restored
     * @param txn the transaction to be applied
     */
    public void processTransaction(TxnHeader hdr,DataTree dt,
            Map<Long, Integer> sessions, Record txn)
        throws KeeperException.NoNodeException {
        ProcessTxnResult rc;
        switch (hdr.getType()) {
        case OpCode.createSession:
            sessions.put(hdr.getClientId(),
                    ((CreateSessionTxn) txn).getTimeOut());
            if (LOG.isTraceEnabled()) {
                ZooTrace.logTraceMessage(LOG,ZooTrace.SESSION_TRACE_MASK,
                        "playLog --- create session in log: 0x"
                                + Long.toHexString(hdr.getClientId())
                                + " with timeout: "
                                + ((CreateSessionTxn) txn).getTimeOut());
            }
            // give dataTree a chance to sync its lastProcessedZxid
            rc = dt.processTxn(hdr, txn);
            break;
        case OpCode.closeSession:
            sessions.remove(hdr.getClientId());
            if (LOG.isTraceEnabled()) {
                ZooTrace.logTraceMessage(LOG,ZooTrace.SESSION_TRACE_MASK,
                        "playLog --- close session in log: 0x"
                                + Long.toHexString(hdr.getClientId()));
            }
            rc = dt.processTxn(hdr, txn);
            break;
        default:
            rc = dt.processTxn(hdr, txn);
        }

        /**
         * This should never happen. A NONODE can never show up in the 
         * transaction logs. This is more indicative of a corrupt transaction
         * log. Refer ZOOKEEPER-1333 for more info.
         */
        if (rc.err != Code.OK.intValue()) {          
            if (hdr.getType() == OpCode.create && rc.err == Code.NONODE.intValue()) {
                int lastSlash = rc.path.lastIndexOf('/');
                String parentName = rc.path.substring(0, lastSlash);
                LOG.error("Parent {} missing for {}", parentName, rc.path);
                throw new KeeperException.NoNodeException(parentName);
            } else {
                LOG.debug("Ignoring processTxn failure hdr: " + hdr.getType() +
                        " : error: " + rc.err);
            }
        }
    }
    
    /**
     * the last logged zxid on the transaction logs
     * @return the last logged zxid
     */
    public long getLastLoggedZxid() {
        FileTxnLog txnLog = new FileTxnLog(dataDir);
        return txnLog.getLastLoggedZxid();
    }

    /**
     * save the datatree and the sessions into a snapshot
     * @param dataTree the datatree to be serialized onto disk
     * @param sessionsWithTimeouts the sesssion timeouts to be
     * serialized onto disk
     * @throws IOException
     */
    public void save(DataTree dataTree,
            ConcurrentHashMap<Long, Integer> sessionsWithTimeouts)
        throws IOException {
        long lastZxid = dataTree.lastProcessedZxid;
        File snapshotFile = new File(snapDir, Util.makeSnapshotName(lastZxid));
        LOG.info("Snapshotting: 0x{} to {}", Long.toHexString(lastZxid),
                snapshotFile);
        snapLog.serialize(dataTree, sessionsWithTimeouts, snapshotFile);
        
    }

    /**
     * truncate the transaction logs the zxid
     * specified
     * @param zxid the zxid to truncate the logs to
     * @return true if able to truncate the log, false if not
     * @throws IOException
     */
    public boolean truncateLog(long zxid) throws IOException {
        // close the existing txnLog and snapLog
        close();

        // truncate it
        FileTxnLog truncLog = new FileTxnLog(dataDir);
        boolean truncated = truncLog.truncate(zxid);
        truncLog.close();

        // re-open the txnLog and snapLog
        // I'd rather just close/reopen this object itself, however that 
        // would have a big impact outside ZKDatabase as there are other
        // objects holding a reference to this object.
        txnLog = new FileTxnLog(dataDir);
        snapLog = new FileSnap(snapDir);

        return truncated;
    }
    
    /**
     * the most recent snapshot in the snapshot
     * directory
     * @return the file that contains the most 
     * recent snapshot
     * @throws IOException
     */
    public File findMostRecentSnapshot() throws IOException {
        FileSnap snaplog = new FileSnap(snapDir);
        return snaplog.findMostRecentSnapshot();
    }
    
    /**
     * the n most recent snapshots
     * @param n the number of recent snapshots
     * @return the list of n most recent snapshots, with
     * the most recent in front
     * @throws IOException
     */
    public List<File> findNRecentSnapshots(int n) throws IOException {
        FileSnap snaplog = new FileSnap(snapDir);
        return snaplog.findNRecentSnapshots(n);
    }

    /**
     * get the snapshot logs that are greater than
     * the given zxid 
     * @param zxid the zxid that contains logs greater than 
     * zxid
     * @return
     */
    public File[] getSnapshotLogs(long zxid) {
        return FileTxnLog.getLogFiles(dataDir.listFiles(), zxid);
    }

    /**
     * append the request to the transaction logs
     * @param si the request to be appended
     * returns true iff something appended, otw false 
     * @throws IOException
     */
    public boolean append(Request si) throws IOException {
        return txnLog.append(si.hdr, si.txn);
    }

    /**
     * commit the transaction of logs
     * @throws IOException
     */
    public void commit() throws IOException {
        txnLog.commit();
    }

    /**
     * roll the transaction logs
     * @throws IOException 
     */
    public void rollLog() throws IOException {
        txnLog.rollLog();
    }
    
    /**
     * close the transaction log files
     * @throws IOException
     */
    public void close() throws IOException {
        txnLog.close();
        snapLog.close();
    }
}

File: zookeeper/server/persistence/FileTxnSnapLog.cpp
/*
 * FileTxnSnapLog.cpp
 *
 *  Created on: 2017-11-22
 *      Author: cxxjava@163.com
 */

#include "./FileTxnSnapLog.hh"
#include "./FileTxnLog.hh"
#include "./FileSnap.hh"

namespace efc {
namespace ezk {

sp<ELogger> FileTxnSnapLog::LOG = ELoggerManager::getLogger("FileTxnSnapLog");

FileTxnSnapLog::FileTxnSnapLog(EFile* dataDir, EFile* snapDir) THROWS(EIOException) :
	dataDir_(dataDir, (EString(version) + VERSION).c_str()),
	snapDir_(snapDir, (EString(version) + VERSION).c_str()) {
	LOG->debug(EString::formatOf("Opening datadir:%s snapDir:%s", dataDir->toString().c_str(), snapDir->toString().c_str()));

	if (!dataDir_.exists()) {
		if (!dataDir_.mkdirs()) {
			throw EIOException(__FILE__, __LINE__, ("Unable to create data directory " + dataDir_.toString()).c_str());
		}
	}

	if (!dataDir_.canWrite()) {
		throw EIOException(__FILE__, __LINE__, ("Cannot write to data directory " + dataDir_.toString()).c_str());
	}

	if (!snapDir_.exists()) {
		if (!snapDir_.mkdirs()) {
			throw EIOException(__FILE__, __LINE__, ("Unable to create snap directory " + snapDir_.toString()).c_str());
		}
	}

	if (!snapDir_.canWrite()) {
		throw EIOException(__FILE__, __LINE__, ("Cannot write to snap directory " + snapDir_.toString()).c_str());
	}

	txnLog = new FileTxnLog(&dataDir_);
	snapLog = new FileSnap(&snapDir_);
}

llong FileTxnSnapLog::restore(sp<DataTree> dt, EConcurrentHashMap<llong, EInteger>* sessions,
		PlayBackListener* listener) THROWS(EIOException) {
	snapLog->deserialize(dt, sessions);
	FileTxnLog txnLog(&dataDir_);// = new FileTxnLog(dataDir);
	sp<TxnLog::TxnIterator> itr = txnLog.read(dt->lastProcessedZxid+1);
	llong highestZxid = dt->lastProcessedZxid;
	sp<TxnHeader> hdr;
	ON_FINALLY_NOTHROW(
		if (itr != null) {
			itr->close();
		}
	) {
		while (true) {
			// iterator points to
			// the first valid txn when initialized
			hdr = itr->getHeader();
			if (hdr == null) {
				//empty logs
				return dt->lastProcessedZxid;
			}
			if (hdr->getZxid() < highestZxid && highestZxid != 0) {
				LOG->error(EString::formatOf("%ld(higestZxid) > %ld(next log) for type %d",
						highestZxid, hdr->getZxid(), hdr->getType()));
			} else {
				highestZxid = hdr->getZxid();
			}
			try {
				processTransaction(hdr.get(), dt.get(), sessions, itr->getTxn().get());
			} catch (NoNodeException& e) {
			   throw EIOException(__FILE__, __LINE__, (EString("Failed to process transaction type: ") +
					 hdr->getType() + " error: " + e.getMessage()).c_str(), &e);
			}
			listener->onTxnLoaded(hdr, itr->getTxn());
			if (!itr->next())
				break;
		}
	}}
	return highestZxid;
}

void FileTxnSnapLog::processTransaction(TxnHeader* hdr, DataTree* dt,
		EConcurrentHashMap<llong, EInteger>* sessions, ERecord* txn)
		THROWS(NoNodeException) {
	sp<DataTree::ProcessTxnResult> rc;
	CreateSessionTxn* cstxn = dynamic_cast<CreateSessionTxn*>(txn);
	switch (hdr->getType()) {
	case ZooDefs::OpCode::createSession: {
		sessions->put(hdr->getClientId(),
				new EInteger(cstxn->getTimeOut()));
		if (LOG->isTraceEnabled()) {
			ZooTrace::logTraceMessage(LOG, ZooTrace::SESSION_TRACE_MASK,
					"playLog --- create session in log: 0x"
							+ ELLong::toHexString(hdr->getClientId())
							+ " with timeout: "
							+ cstxn->getTimeOut());
		}
		// give dataTree a chance to sync its lastProcessedZxid
		rc = dt->processTxn(hdr, txn);
		break;
	}
	case ZooDefs::OpCode::closeSession: {
		sessions->remove(hdr->getClientId());
		if (LOG->isTraceEnabled()) {
			ZooTrace::logTraceMessage(LOG, ZooTrace::SESSION_TRACE_MASK,
					"playLog --- close session in log: 0x"
							+ ELLong::toHexString(hdr->getClientId()));
		}
		rc = dt->processTxn(hdr, txn);
		break;
	}
	default:
		rc = dt->processTxn(hdr, txn);
		break;
	}

	/**
	 * Snapshots are lazily created. So when a snapshot is in progress,
	 * there is a chance for later transactions to make into the
	 * snapshot. Then when the snapshot is restored, NONODE/NODEEXISTS
	 * errors could occur. It should be safe to ignore these.
	 */
	if (rc->err != KeeperException::Code::OK) {
		LOG->debug((EString("Ignoring processTxn failure hdr:") + hdr->getType()
				+ ", error: " + rc->err + ", path: " + rc->path).c_str());
	}
}

llong FileTxnSnapLog::getLastLoggedZxid() {
	FileTxnLog txnLog(&dataDir_);
	return txnLog.getLastLoggedZxid();
}

void FileTxnSnapLog::save(sp<DataTree> dataTree,
		EConcurrentHashMap<llong, EInteger>* sessionsWithTimeouts)
		THROWS(EIOException) {
	llong lastZxid = dataTree->lastProcessedZxid;
	EFile snapshotFile(&snapDir_, Util::makeSnapshotName(lastZxid).c_str());
	LOG->info(EString::formatOf("Snapshotting: 0x%s to %s", ELLong::toHexString(lastZxid).c_str(),
			snapshotFile.toString().c_str()).c_str());
	snapLog->serialize(dataTree, sessionsWithTimeouts, &snapshotFile);
}

boolean FileTxnSnapLog::truncateLog(llong zxid) THROWS(EIOException) {
	// close the existing txnLog and snapLog
	close();

	// truncate it
	FileTxnLog truncLog(&dataDir_);
	boolean truncated = truncLog.truncate(zxid);
	truncLog.close();

	// re-open the txnLog and snapLog
	// I'd rather just close/reopen this object itself, however that
	// would have a big impact outside ZKDatabase as there are other
	// objects holding a reference to this object.
	txnLog = new FileTxnLog(&dataDir_);
	snapLog = new FileSnap(&snapDir_);

	return truncated;
}

sp<EFile> FileTxnSnapLog::findMostRecentSnapshot() THROWS(EIOException) {
	FileSnap snaplog(&snapDir_);
	return snaplog.findMostRecentSnapshot();
}

sp<EList<EFile*> > FileTxnSnapLog::findNRecentSnapshots(int n) THROWS(EIOException) {
	FileSnap snaplog(&snapDir_);
	return snaplog.findNRecentSnapshots(n);
}

sp<EArray<EFile*> > FileTxnSnapLog::getSnapshotLogs(llong zxid) {
	EArray<EFile*> files = dataDir_.listFiles();
	return FileTxnLog::getLogFiles(&files, zxid);
}

boolean FileTxnSnapLog::append(sp<Request> si) THROWS(EIOException) {
	return txnLog->append(si->hdr.get(), si->txn.get());
}

} /* namespace ezk */
} /* namespace efc */
